sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -0,0 +1,435 @@
+        String cryptoAlias = CRYPTO_KEY_PREFIX + index;
+        String authAlias = AUTH_KEY_PREFIX + index;
+        KeyMetaDataStore.KeyMetaDataEntry cryptoEntry = keyMetaDataStore.getEntry(cryptoAlias);
+        KeyMetaDataStore.KeyMetaDataEntry authEntry = keyMetaDataStore.getEntry(authAlias);
+        SecretKey cryptoKey = recoverSecretKey(cryptoAlias, cryptoEntry);
+        SecretKey authKey = recoverSecretKey(authAlias, authEntry);
+        return new Pair<>(cryptoKey, authKey);
+    }
+
+    public Pair<SecretKey, SecretKey> getLastPair() {
+        int lastIndex = getNumOfPairs() - 1;
+        return getOnePair(lastIndex);
+    }
+
+    //Return all pair of keys as a stack. The most recent pair of keys are to be popped first
+    public Queue<Pair<SecretKey, SecretKey>> getAllPairs() {
+        synchronized (RegistrationKeyManager.class) {
+            int pairs = getNumOfPairs() - 1;
+            Queue<Pair<SecretKey, SecretKey>> res = new LinkedList<>();
+            for (int i = pairs; i >= 0; i--) {
+                Pair<SecretKey, SecretKey> pair = getOnePair(i);
+                res.offer(pair);
+            }
+            return res;
+        }
+    }
+
+    private SecretKey secretKeyFromStr(String str) {
+        byte[] bytes = Base64Util.decodeString(str);
+        return new SecretKeySpec(bytes, 0, bytes.length, "AES");
+    }
+
+    class EnCryptor {
+        private byte[] encryption;
+        private byte[] iv;
+
+        private static final String TRANSFORMATION_SYMMETRIC = "AES/GCM/NoPadding";
+
+        EnCryptor() {
+        }
+
+        void encryptText(final Key encryptionKey, final String textToEncrypt) {
+            final Cipher cipher;
+            try {
+                cipher = Cipher.getInstance(TRANSFORMATION_SYMMETRIC);
+                cipher.init(Cipher.ENCRYPT_MODE, encryptionKey);
+                iv = cipher.getIV();
+                encryption = cipher.doFinal(textToEncrypt.getBytes("UTF-8"));
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | UnsupportedEncodingException e) {
+                throw clientLogger.logExceptionAsError(new RuntimeException("Failed to encrypt data", e));
+            } catch (IllegalBlockSizeException e) {
+                throw clientLogger.logExceptionAsError(new RuntimeException("Failed to encrypt data", e));
+            } catch (BadPaddingException e) {
+                throw clientLogger.logExceptionAsError(new RuntimeException("Failed to encrypt data", e));
+            } catch (InvalidKeyException e) {
+                throw clientLogger.logExceptionAsError(new RuntimeException("Failed to encrypt data", e));
+            }
+        }
+    }
+
+    class DeCryptor {
+        private static final String TRANSFORMATION_SYMMETRIC = "AES/GCM/NoPadding";
+
+        DeCryptor() {
+        }
+
+        String decryptData(final Key decryptionKey, final byte[] encryptedData, final byte[] encryptionIv) {
```

Comment:
nit; In cryptography, encrypted data is typically called `ciphertext`.
```suggestion
        String decryptData(final Key decryptionKey, final byte[] ciphertext, final byte[] encryptionIv) {
```